### ISSUE TITLE
[7.0] [8.0] Odoo should set the bounce address

### ISSUE THREAD

**Issue by mvaled** [odoo/odoo#3347]
#### Summary:

When sending notifications because an external email has been received, Odoo identifies itself with the SMTP server using the original email's From header.  This might cause the MTA be considered a spammer when relaying emails.

Note: I'm using the phrase «identifies with the SMTP server» not to say it authenticates with the server, but to mean that OpenERP issues a MAIL FROM command with a specific email address.

#### Environment:

I have setup a local Postfix + Dovecot.  OpenERP setting for outgoing email is simply localhost:25 without authentication.  OpenERP's catch-all domain is «oo.example.com» (in OpenERP 8.0 there's also a catch-all alias).

Postfix relays to OpenERP's mailgate script all addresses for the domain «oo.example.com».   Via its postgres integration it accepts emails if there's a configured alias in OpenERP.

Postfix accepts as local addresses the domains «external.example.com», «other.example.com».  Any email to address in these domains are accepted.

When postfix delivers emails to dovecot, it places them to a single catch-all inbox regardless of the recipient address, so I can read all emails in a single IMAP account.

#### Steps to reproduce

Log into a new DB as admin and install the Messaging application.  You may need to install HR in Odoo 8.0 to have the «Share with my followers» box in your Messaging Inbox (see issue #2042).

Send a single email to «bob@external.example.com», and «alice@other.example.com». 

From an email application connect to your local IMAP server (dovecot) and check you have the two copies (one for each recipient) delivered.  *Reply* to «admin@oo.example.com» identifying yourself as bob with the SMTP server.  

Postfix will then relay the message to OpenERP, and since it gathers this email as a reply to an existing mail thread it will notify alice with the response from bob.  In doing so it will also identify itself to the SMTP server as bob@external.example.com.

**Comment by thomich** [comment]
is it correct, that this is still no pushed/fixed?

**Comment by mvaled** [comment]
Hi Thomas,

I stopped pushing for this a while ago. As you can see in the issue
history there are many rebased commits in the associated pull request. 

Instead I created a fork which I'm using so far on our servers
(https://github.com/merchise-autrement/odoo/tree/merchise-7.0).  This
fork removed the problem that some our partners with Gmail-provided
addresses were not receiving some emails. 

The only issue I have detected is that some external parties (mostly
those that use MS products) are getting confused about the Sender header
and assume that they can write to the postmaster, instead of replying to
the Reply-To address.

I'm doing a new release that does not set the Sender while keeping the
Return-Path, to avoid that.  Actually Odoo's mailing list seem to use a
Variable Envelop Return-Path like
"<bounce-2156894-mail.group-62@mail.odoo.com>" but don't use the Sender
header.

Best regards,
Manuel.

**Comment by thomich** [comment]
exact the problem with the random generated senders address is the issue...
if you use domain1.com as alias domain.. an setup outgoing server which is mail.domain1.com and you send any message from odoo (project, issue,...), mail.domain1.com will response with:
```
MailDeliveryException: (u'Mail Delivery Failed', u"Mail delivery failed via SMTP server 'mail.domain1.com'.\nSMTPSenderRefused: 553\n5.3.0 <bounce-75-mail.group-3@domain1.com>... No such user here\nbounce-75-mail.group-3@domain1.com")
```

user that creates the message has a correct sending email address like: user@domain1.com, which also exist on mail.domain1.com...

facts: odoo should use users email address for sender (from) parameter AND respect also mail template... the we configured the senders address, but not working...

**Comment by thomich** [comment]
workaround: remove system parameter mail.bounce.alias

**Comment by thomich** [comment]
you referenced fix wil fix Return-Path, but there is another method, that is also set the Return-Path to the bounce.alias, check /server/addons/mail/mail_mail.py line 256+
```
# specific behavior to customize the send email for notified partners
email_list = []
if mail.email_to:
    email_list.append(self.send_get_email_dict(cr, uid, mail, context=context))
for partner in mail.recipient_ids:
    email_list.append(self.send_get_email_dict(cr, uid, mail, partner=partner, context=context))
# headers
headers = {}
bounce_alias = self.pool['ir.config_parameter'].get_param(cr, uid, "mail.bounce.alias", context=context)
catchall_domain = self.pool['ir.config_parameter'].get_param(cr, uid, "mail.catchall.domain", context=context)
if bounce_alias and catchall_domain:
    if mail.model and mail.res_id:
        headers['Return-Path'] = '%s-%d-%s-%d@%s' % (bounce_alias, mail.id, mail.model, mail.res_id, catchall_domain)
    else:
        headers['Return-Path'] = '%s-%d@%s' % (bounce_alias, mail.id, catchall_domain)
if mail.headers:
    try:
        headers.update(eval(mail.headers))
    except Exception:
        pass

# Writing on the mail object may fail (e.g. lock on user) which
# would trigger a rollback *after* actually sending the email.
# To avoid sending twice the same email, provoke the failure earlier
mail.write({'state': 'exception'})
mail_sent = False
```

**Comment by mvaled** [comment]
Hi Thomas,

I had a hard time trying to understand your problem.  A few minutes ago @odony landed my patch in the odoo-dev repo.

Why do think this is a problem.  Notice the From is never being changed.

**Comment by mvaled** [comment]
Furthermore, this patch is for the 7.0 branch.  I haven't migrated to 8.0.

**Comment by thomich** [comment]
just to inform... the mail.bounce.alias is new in Odoo 8... i don't see that in config or in source of Odoo 7...

**Comment by mvaled** [comment]
That's why we fallback to "postmaster-odoo".  The patch was made so that
it works in Odoo 7.0 but to be easily ported to Odoo.

**Comment by l-arnold** [comment]
Question about Bounce Alias.  How do we review the bounces?  Should there be a Partner or User who's alias is "Bounce".  Would be good to know when a Bounce occurs.  Likely programatic but not clear from here.  I did rename my "catchall" address to "erp" which seems to look better on "reply to".

**Comment by mvaled** [comment]
Hi @odony,

Since a few days we're getting bounces from the hotmail server like this:

    Mar 25 07:48:52 mail postfix/smtp[29382]: AB687428625: to=<XXXXXX@hotmail.fr>,
    relay=mx3.hotmail.com[65.55.92.184]:25, delay=1.3, delays=0.01/0.01/0.34/0.97,
    dsn=5.7.0, status=bounced (host mx3.hotmail.com[65.55.92.184] said: 550
    5.7.0 (SNT004-MC4F16) Unfortunately, messages from (85.XXX.XXX.XXX) on behalf
    of (yahoo.com) could not be delivered due to domain owner policy
    restrictions. (in reply to end of DATA command))

It seem that hotmail are now also checking the headers not just the envelop address to forbid issuing a email.  Have notice something similar lately in your servers? Or simply hotmail.fr is broken?

**Comment by mvaled** [comment]
BTW, I don't know why this is still open.

**Comment by mvaled** [comment]
I've just registered myself in hotmail.com and subscribed to the Odoo Framework mailing list.

**Comment by odony** [comment]
I've seen a couple dozens such delivery failures in the last week, most of them 
for messages coming from twitter.com, paypal.de, facebookmail.com, and a few 
for yahoo.com like you.

Hotmail's SMTP stack seems to implement a stricter SPF check, not only applying 
it to the From/Sender header fields (which seems wrong), but also not 
respecting the SPF semantics. Twitter.com does have a hardfail SPF record 
(-all), but Yahoo.com's one is neutral (?all) as far as I can see, so it should 
not impact the delivery policy in case of failed check.

That might prove to be a problem for hotmail users who'd like to subscribe to 
mailing-lists, as they'll miss random emails depending on who's sending them.

**Comment by mvaled** [comment]
Le 27/03/15 08:00, Olivier Dony a écrit :
> That might prove to be a problem for hotmail users who'd like to
> subscribe to
> mailing-lists, as they'll miss random emails depending on who's sending
> them.

I did finally received a couple of emails from the Odoo mailing list in
my hotmail.com inbox.  But 'only' a couple of hours after I sent them.

I also subscribed to nginx's mailing list but not even the confirmation
has reached my inbox yet.

So I assume this has to do with both a very wrong mail infrastructure in
hotmail's servers, and also a kind of very sloooow SPAM/SPF check with
wrong semantics.

**Comment by mvaled** [comment]
Hi @odony,

Definitely this lately bounces are related to the "new" DMARC policy of Yahoo.com.  I expect some other providers will follow.

Are you having any kind of public debate about how to solve this issue in Odoo? (So that I can go there instead of polluting this PR)

**Comment by odony** [comment]
@mvaled I'm not aware of any public discussion about this topic, but you're welcome to start a thread on the [community list](https://www.odoo.com/groups/59) about it. Just don't post from a Yahoo email ;-)
I guess a simple option would be a global setting to force rewriting the `From` header for any outgoing email (similar to mailman's `from_is_list`) setting. One problem is that there won't be any obvious place where recipients can see the email address of the Sender, since the Reply-To is already supposed to be the catchall/list address. We can preserve the name of the original `From`, but keeping the email address would make it quite ugly, and I don't see any appropriate header to put it.
E.g: 

        From: "Jane Doer via Odoo" <catchall@mail.odoo.com>
or 

        From: "Jane Doer (jane@yahoo.com) via Odoo" <catchall@mail.odoo.com>
instead of

        From: "Jane Doer" <jane@yahoo.com>

**Comment by yelizariev** [comment]
similar bug report https://github.com/odoo/odoo/issues/5864 about From field

My opion about From header. It should be just

    catchall@mail.odoo.com

and maybe From value should be added in Body.

I think a lot of mail service save name of recipient by email, so some mess can happens:

1. User receives email 

    From: "Jane Doer via Odoo" <catchall@mail.odoo.com>

2. Mail service or mail client save  "Jane Doer via Odoo" as a name for catchall@mail.odoo.com

3. User  receives another email 

    From: "Another Guy via Odoo" <catchall@mail.odoo.com>

4. When user try to reply to one of above emails wrong name of recepient can be used by mail service.
OR another situation: user wants to send email to Jane Doer and mail service suggest to use

    "Jane Doer via Odoo" <catchall@mail.odoo.com>

(catchall@mail.odoo.com can be hidden).

**Comment by l-arnold** [comment]
You can also changevthe "catchall"  name within settings.  Look for Catchall somewhere there in Odoo settings changed mine to "erp" and it seems tp work for folks.  It is a bit odd how the system works but quite geniotic too.

**Comment by jdegoy** [comment]
I consider this bug as not resolved as long as e-mails coming out of Odoo are being classified as Junk/Spam. All MS Exchange servers are classifying mails as Junk when the triangulation is being used (via catchall mailbox). Even if you say it is because MS, ... is being more stringent than they should, we won't convince them to apply a different filtering strategy...
So what can we do with standard Odoo to fix this?

**Comment by odony** [comment]
@jdegoy Most email systems (including MS Exchange) includes numerous spam detection heuristics and combine the results to decide upon the classification. Any one of these heuristics (or a combination) may cause your emails to be classified as junk , so your first step to solve the problem is to find out *what* is causing the messages to be considered junk.
Refer to the manual, most classifiers include the results of their tests in dedicated email headers in order to help troubleshooting false positives.
Analyzing email delivery problems is complex and requires gathering accurate as much accurate details as possible, including headers and log for every step of the delivery, from sender to final recipient.
Unless you investigate in that direction and provide all the details, there's little we can do to help you. You should probably open a different issue as well, instead of continuing this one.

BTW the original author of this issue closed it themselves, so we can safely consider *this particular issue* closed.

**Comment by odony** [comment]
For the record, SPF protection was *by design* only meant to verify the `Return-Path`/`Sender`/`Bounce`/`Envelope-from` of an email (at the SMTP transaction level), and by no means the "From" header of an email. This is explained in the [SPF FAQ](http://www.openspf.org/FAQ/Envelope_from_scope).
That's why this issue was closed with the addition of a `mail.catchall.alias` system parameter that lets you override this default field. It doesn't change the behavior for the `From` header in any manner.

**Comment by mvaled** [comment]
@jdegoy I agree with @odony.  This issue should remain closed. We're happily exchanging emails with MS users and they haven't complained about our emails being miss-classified as Junk.

**Comment by lasley** [comment]
Rare case having the `-all` in SPF, but this is something that I do for all domains I manage and is very much a problem for me. 

I can guarantee that many services including Gmail do validate SPF on the FROM header and will hard fail in the event of a mismatch, as indicated in the below headers:

```
Authentication-Results: mx.google.com;
       spf=pass (google.com: domain of postmaster-odoo@odoo-community.org designates 2a01:4f8:a0:430d::2 as permitted sender) smtp.mailfrom=postmaster-odoo@odoo-community.org;
       dmarc=fail (p=QUARANTINE sp=QUARANTINE dis=QUARANTINE) header.from=laslabs.com
```

I am testing a solution in OCA/social#195

**Comment by odony** [comment]
@lasley Indeed, DMARC modifies the landscape. It does not simply require passing SPF+DKIM checks, but it requires those checks to be _aligned_ with the domain of the RFC5322 _From header_.  This means you're forced to have matching _From:_ and _Envelope-From/Return-Path_ domains (unlike pure SPF or DKIM checks).
The consequence is that having a locked down DMARC policy **and** participating in mailing-lists is close to impossible at the moment. I'd recommend reading this [3 part article](https://blogs.msdn.microsoft.com/tzink/2015/05/28/solving-the-problem-of-dmarcs-incompatibility-with-mailing-lists-part-1/), which nicely describes the situation, and why the possible workarounds are broken in different ways:
 - adding the author's email in a "Sender:" header is doubly broken: most email clients ignore it entirely (the recipient will never know the author!), and RFC5322 intended this header to indicate the transmitter, not the author, so when Outlook displays it, it's shown backwards
 - adding the author's name in the "From" header as "Author via Odoo <...>" (as suggested in [this earlier comment](https://github.com/odoo/odoo/issues/3347#issuecomment-90486636)) has the downside of messing up address books, as @yelizariev mentioned

That last option might still a sensible workaround, if we make it an opt-in setting that needs to be manually enabled using a system parameter. I'd be happy to review it if anyone wants to make a PR, even in 10.0 (this could be considered a fix for strict DMARC cases, and would need an explicit system parameter to be activated anyways).

@yelizariev suggested we put the author's info in the body. That would be another option, but in order not to confuse the recipient, the email then needs to look like a system notification telling you that someone sent a message. This works well for simple posts on documents (that already look like notifications in Odoo 10), but would certainly be strange for Odoo mailing-list channels, like the Community list. I can only imagine the mess in quoted replies.

I guess we don't have any perfect solution yet...

**Comment by odony** [comment]
Some more thoughts on the possible workaround:
- Rewriting the `From` could be made dynamic depending on the DMARC policy of the recipient. It seems this is the way [Threadable solved the issue](http://blog.threadable.com/how-threadable-solved-the-dmarc-problem). The downside is that it requires an extra DNS lookup for every recipient, but it means only people who really need to see the altered mail would get it. And we would only need to do the lookup for forwarded emails.
- Perhaps the system parameter toggle could also include a whitelist of domains for which no rewriting should take place even if there is a DMARC policy (because the outgoing SMTP server is known to be allowed by the DMARC policy)

**Comment by lasley** [comment]
@odony - Thanks for the detailed elaboration. I agree 100%, and am happy to hear you will accept a PR on this. 

I agree that the manipulation of the FROM header is probably the way to go. This is the way that JIRA, GitLab, and a bunch of other products handle it.

The Threadable approach is interesting. I think I like it a lot - I wonder if we can get something similar to play nicely in Odoo. I don't think the DNS query is too much overhead given the benefits.

Whitelist is totally sensibly and easy to implement. Will do 👍 

This is a side project for me so not actually on the project calendar. That said, shouldn't be more than a few weeks.

**Comment by gdgellatly** [comment]
@lasley @odony One thing to consider, and a real life example just popped up on Odoo mailing list, is when a mailing list user (eg. me) finally decides its time to block rants from another user (e.g. not me).  Currently any rewriting of the From header would regress that currently very useful feature.

**Comment by lasley** [comment]
@gdgellatly - I think this is one of those can't have your cake and eat it situations honestly. To counter the inability to block by email address, I advise that Odoo is incompatible with a significant chunk of mail relay services - such as SES. This isn't just a DEMARC issue IMO.

I'm curious how your blocks are working though? Spam heuristics should be smarter than the simple FROM header, so I'm assuming that you're using a manual rule? Why not block the canonical name in the FROM instead of the email?

**Comment by gdgellatly** [comment]
@lasley it's just my personal Gmail account. It's filters don't have that
option.

**Comment by lasley** [comment]
@gdgellatly - You do have that option in gmail:

```
from:"Dave Lasley" 
```
would filter on the "Dave Lasley" canonical name in the from header, in the same way that:

```
from:"dave@laslabs.com"
```

would match for my email

**Comment by mvaled** [comment]
We had to take ownership of the mail long ago... The reason was quite different:  There are so many broken mail clients out there.  Many didn't respect the Reply-To, and an awful lot didn't bother to include the Message-Id in the References (nor In-Reply-To) when replying.  This last bit implied that replies wouldn't be placed in their proper thread.   

We used a custom MailTransport and MailRouter (https://github.com/merchise/xopgi.mail_threads#usage) that replaced From and created a unique address per thread (e.g 'mailbox+jd8d7f8778ccf').  For this to work properly in Odoo 8, we had to modify the 'mail' addon: https://github.com/merchise-autrement/odoo/blob/merchise-8.0/addons/mail/__init__.py#L30
